### PR TITLE
[Frontend][Core] Add plumbing to support audio language models

### LINKED ISSUE #7335: [Feature]: support voice llm like cosyvoice
### 🚀 The feature, motivation and pitch

As we can see that vllm already support many vision language models. Do we have a plan to support voice models like cosyVoice?

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by petersalas**
This adds infrastructure to support multi-modal audio language models.

- `MultiModalDataBuiltins` defines audio as a tuple of a `numpy.ndarray` and the sample rate. (`librosa` is used to parse the audio.)
- `DEFAULT_PLUGINS` now includes an `AudioPlugin`
- The OpenAI API frontend tentatively supports `"audio_url"` content parts, albeit with the same restrictions that currently apply to `"image_url"` parts (e.g. only one multi-modal chunk per prompt).
- A number of methods/types that were only nominally vision-specific are renamed to be more generically multi-modal. For example, `SupportsVision` is now `SupportsMultiModal`.

This PR does not include any such audio models, but does include a test that exercises the OpenAI frontend and audio pipeline (see `tests/entrypoints/openai/test_audio.py`).

FIX #7335

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: docs/source/models/vlm.rst [hunk 1]
```diff
@@ -150,7 +150,7 @@ A full code example can be found in `examples/openai_vision_api_client.py <https
 
     .. code-block:: shell
 
-        export VLLM_IMAGE_FETCH_TIMEOUT=<timeout>
+        export VLLM_MULTIMODAL_FETCH_TIMEOUT=<timeout>
```
Comment (DarkLight1337) [inline docs/source/models/vlm.rst:153]: Given that different types of data may have different sizes (think image vs video), it may be better to have a separate timeout for each modality.

File: vllm/entrypoints/chat_utils.py [hunk 2]
```diff
@@ -21,12 +22,25 @@
...
+from vllm.multimodal.utils import (async_get_and_parse_audio,
+                                   async_get_and_parse_image)
 from vllm.transformers_utils.tokenizer import AnyTokenizer
 
 logger = init_logger(__name__)
 
 
+class AudioURL(TypedDict, total=False):
+    url: Required[str]
+    """Either a URL of the audio or the base64 encoded image data."""
```
Comment (DarkLight1337) [inline vllm/entrypoints/chat_utils.py:34]:
```suggestion
    """Either a URL of the audio or the base64 encoded audio data."""
```

File: docs/source/models/vlm.rst (hunk 1 above)
Comment (DarkLight1337) [inline docs/source/models/vlm.rst:153]: Given that different types of data may have different sizes (think image vs video), it may be better to have a separate timeout for each modality.

File: vllm/model_executor/models/internvl.py [hunk 3]
```diff
@@ -450,10 +450,10 @@ def forward(
         if image_input is not None:
             inputs_embeds = self.language_model.model.get_input_embeddings(
                 input_ids)
-            vision_embeddings = self._process_image_input(image_input)
-            inputs_embeds = merge_vision_embeddings(input_ids, inputs_embeds,
-                                                    vision_embeddings,
-                                                    self.img_context_token_id)
+            vit_embeds = self.extract_feature(image_input["data"])
```
Comment (DarkLight1337) [inline vllm/model_executor/models/internvl.py:453]: Is this change intended?

File: vllm/multimodal/utils.py [hunk 4]
```diff
@@ -1,11 +1,14 @@
 import base64
 from io import BytesIO
-from typing import Union
+from typing import Tuple, Union
 
+import librosa
+import numpy as np
+import soundfile
```
Comment (DarkLight1337) [inline vllm/multimodal/utils.py:7]: Should we add `soundfile` to `requirements-common.txt` as well?

File: tests/entrypoints/openai/test_audio.py [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+                raise RuntimeError("Server exited unexpectedly.") from err
+
+            time.sleep(0.5)
+            if time.time() - start > MAX_SERVER_START_WAIT_S:
+                raise RuntimeError("Server failed to start in time.") from err
+
+    try:
+        yield client
+    finally:
+        server.kill()
```
Comment (DarkLight1337) [inline tests/entrypoints/openai/test_audio.py:131]: Note to future self: Use `RemoteOpenAIServer` like the other tests once we support at least one audio model.

**Review by DarkLight1337 — COMMENTED** [review]
Thanks for implementing this! Overall it looks good, just some small things that need to be resolved.

**Comment by DarkLight1337** [comment]
Are you going to work on adding a new audio model after this PR?

**Comment by ywang96** [comment]
This is exciting! Thank you for kicking off the initiative to support audio LMMs!

File: docs/source/models/vlm.rst (hunk 1 above)
Comment (petersalas) [reply docs/source/models/vlm.rst]: Ah, good thought. I split it out accordingly.

**Review by DarkLight1337 — APPROVED** [review]

**Comment by petersalas** [comment]
> Are you going to work on adding a new audio model after this PR?

If you're open to it, I'd love to bring https://github.com/fixie-ai/ultravox into vLLM. We're currently serving it via an externally registered model + some monkeypatches to handle in audio via `"image_url"` but we'd certainly rather it be directly supported!

**Comment by DarkLight1337** [comment]
Going to merge this so I can incorporate it into my #7126 tomorrow.

File: requirements-common.txt [hunk 6]
```diff
@@ -22,4 +22,6 @@ outlines >= 0.0.43, < 0.1 # Requires torch >= 2.1.0
 typing_extensions >= 4.10
 filelock >= 3.10.4 # filelock starts to support `mode` argument from 3.10.4
 pyzmq
+librosa # Required for audio processing
+soundfile # Required for audio processing
```
Comment (ywang96) [inline requirements-common.txt:26]: Can you mock these in `vllm/docs/source/conf.py`? I think otherwise it'll break our doc. Thanks!

**Review by ywang96 — CHANGES_REQUESTED** [review]
Just one change needed - Thank you for making this PR! @petersalas

File: requirements-common.txt (hunk 6 above)
Comment (petersalas) [reply requirements-common.txt:26]: No problem -- sorry to miss that!

**Review by ywang96 — APPROVED** [review]
